### PR TITLE
Secant Method Algorithm for Numerical Methods

### PR DESCRIPTION
**Pull request by fkkarakurt**
#### Checklist

- [x] Added description of change
- [x] Added file name matches [File name guidelines](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md#New-File-Name-guidelines)
- [x] Added tests and example, test must pass
- [x] Added documentation so that the program is self-explanatory and educational - [Doxygen guidelines](https://www.doxygen.nl/manual/docblocks.html)
- [x] Relevant documentation/comments is changed or added
- [x] PR title follows semantic [commit guidelines](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md#Commit-Guidelines)
- [x] Search previous suggestions before making a new one, as yours may be a duplicate.
- [x] I acknowledge that all my contributions will be made under the project's license.

Notes:

### CODE REVIEW

File: numerical_methods/secant_method.cpp [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+ *
+ * \author [Fatih Küçükkarakurt](https://github.com/fkkarakurt)
+ */
+
+// Secant Method for the root of f(x)
+// Analytically the solution is x^4-29 = 0
+// => x = 29^(1/4) = 2,3205957871060836757951626724362...
+#include <cmath>
+#include <iomanip>
+#include <iostream>
```
Comment (Panquesito7) [inline numerical_methods/secant_method.cpp:19]: Please add a one-line description of what the library/header is for (**see the example below**).

```c++
#include <iostream>    /// for IO operations
#include <cassert>    /// for assert
```

File: numerical_methods/secant_method.cpp [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+
+double f(double x) { return x * x * x * x - 29.0; } // The Function f(x)
+
+/** Main Function */
```
Comment (Panquesito7) [inline numerical_methods/secant_method.cpp:23]:
```suggestion
/**
 * @brief Main function
 * @returns 0 on exit
 */
```

File: numerical_methods/secant_method.cpp [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+    /** Start Secant Method */
+    do {
+        double error = (x1 - x0) / (f(x1) - f(x0)) * f(x1); // Estimate the error
+        std::cout << "Root = " << x1 << std::setw(5) << " Error = " << error << std::endl;
+
+        if (abs(error) < tolerance)
+            break;         // Terminate if the tolerance is satisfied
+        x0 = x1;         // Improve x0
+        x1 = x1 - error; // Improve x1
+    } while (1);
```
Comment (Panquesito7) [inline numerical_methods/secant_method.cpp:37]:
```suggestion
    } while (1);
    return 0;
```

File: numerical_methods/secant_method.cpp [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+ */
+
+// Secant Method for the root of f(x)
+// Analytically the solution is x^4-29 = 0
+// => x = 29^(1/4) = 2,3205957871060836757951626724362...
+#include <cmath>
+#include <iomanip>
+#include <iostream>
+
+double f(double x) { return x * x * x * x - 29.0; } // The Function f(x)
```
Comment (Panquesito7) [inline numerical_methods/secant_method.cpp:21]: Please add documentation with detailed/brief description, parameters description, etc. etc..
Check our [contributing guidelines](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md). You can see an example of the same here:
```c++
/**
 * Utility function to print matrix
 * @tparam n number of matrix size
 * @param board matrix where numbers are saved
 */
template <size_t n>
void printSolution(const std::array<std::array<int, n>, n> &board) {
    std::cout << "\n";
    for (int i = 0; i < n; i++) {
        for (int j = 0; j < n; j++) {
            std::cout << "" << board[i][j] << " ";
        }
        std::cout << "\n";
    }
}
```

**Review by Panquesito7 — APPROVED** [review]
Great work! 🚀

**Comment by fkkarakurt** [comment]
I edited it a bit more and added annotations.

Have a nice day!

File: numerical_methods/secant_method.cpp [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+* Analytically the solution is
+* \f[
+* x^4-29 = 0
+* x = 29^(1/4) = 2,3205957871060836757951626724362...
+* \f]
+*/
+
+#include <cmath> /// for built-in mathematical functions
+#include <iomanip> /// to identify the manipulator and setprecision
+#include <iostream> /// for IO operations
```
Comment (Panquesito7) [inline numerical_methods/secant_method.cpp:25]:
```suggestion
#include <cmath>    /// for built-in mathematical functions
#include <iomanip>  /// to identify the manipulator and `setprecision`
#include <iostream>  /// for IO operations
```

File: numerical_methods/secant_method.cpp [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+*/
+double f(double x) { return x * x * x * x - 29.0; } // The Function f(x)
+
+/**
+* @brief Main function
+* @returns 0 on exit
+*/
+int main() {
+   /**
+    * @brief Information required for the secant method
```
Comment (Panquesito7) [inline numerical_methods/secant_method.cpp:39]: Seems like the code above could be separated as a function with this documentation, IMO.

File: numerical_methods/secant_method.cpp [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+#include <cmath> /// for built-in mathematical functions
+#include <iomanip> /// to identify the manipulator and setprecision
+#include <iostream> /// for IO operations
+
+/**
+* @brief Definition the f(x) fuction
+* @param x unknown x value
+*/
+double f(double x) { return x * x * x * x - 29.0; } // The Function f(x)
+
```
Comment (Panquesito7) [inline numerical_methods/secant_method.cpp:32]: Please add self-test cases so we can verify your program works properly (**see the [typical structure of a program](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md#typical-structure-of-a-program) as a reference**). Let us know if you need any help here or on our [**Discord server**](https://the-algorithms.com/discord). 😃

**Review by Panquesito7 — CHANGES_REQUESTED** [review]
Great work! 🚀

**Pull request closed without merging**